README.md
# Zway-BaseModule

Module which provides many useful functions for other automation modules.
Has no user facing functionality on its own.

As a user of this app you just have to instantiate it once via Apps >
Local Apps > Base Module > Add App. No configuration settings are required.

# Developers

The following documentation is only relevant to developers of Zway automation
modules.

For basic usage just replace one line from the module initialisation

```javascript
inherits(MyModule, BaseModule); // instead of AutomationModule
```

Also add BaseModule as a dependency in module.json
```json
{
   "moduleName": "MyModule",
   "dependencies" : ["BaseModule"],
   ...
```

And finally instantiate the module via Apps > Local Apps > Base Module >
Add App.

BaseModule helps module authors with these utilities:

* Logging with prefixes instead of console.log and error
* Timestamp and event for real change in every vDev (metrics:changeTime)
* Presence states (via the Presence module)
* Find devices based on various criterias
* Process devices from list of device IDs
* Parse time
* Compare time periods
* Secure timeout handling

# Methods

## log, error

Log messages and errors to console with module name and id prefix
instead of console.log and console.error

## presenceModes

List all available presence modes.

## getPresenceBoolean

Returns presence state as boolean value (true = at home, false = away) from the presence device

## getPresenceMode

Returns the current presence mode as string from the presence device

## getDevices

```javascript
var myBatteryDevices = self.getDevices([['probeTitle','=','Battery']]);
```

Returns a list of devices that match the given criteria. Criteria are
evaluated using the compareDevice method.

## getDevice

Works like getDevices, but returns the first matching device.

## getDeviceValue

```javascript
var temperature = self.getDeviceValue([
    ['probeTitle','=','temperature'],
    ['location','=',3]
],'metrics:level');
```

Returns the selected value of the the first matching device. If no value is
given metrics:level will be returned.
Criteria are evaluated using the compareDevice method

## processDevices

```javascript
self.processDevices([
    ['probeTitle','=','temperature'],
    ['location','=',3]
],function(vDev) { ... });
```

Find the selected devices and process a callback for each device.

## performCommandDevices

Performs a given command on all devices matching the selected criteria.
Criteria are evaluated using the compareDevice method. Additionally the
device auto flag may be set.

The following example turns sets all dimmers in room 3 to 33%, and also sets
the metrics:auto flag to true.

```javascript
var temperature = self.performCommandDevices([
    ['deviceType','=','switchMultilevel'],
    ['location','=',3]
],'exact',{ 'level': 33 },true);
```

## compareDevice

This method is used by performCommandDevices, getDeviceValue, getDevice and
getDevices to check if a device matches given criteria or not.

Criteria must be supplied as an array of arrays where the nested array has
three values.

```
[
    [ KEY, COMPARISON, VALUE ]
]
```

Key may be any key from the virtual device object (with colons when using
nested keys) and additionally zwaveId.

Comparison can be either '=', '!=', '>', '>=', '<' or '<='.

Value can be a string, boolean or array value. If the value is an array, a
device matches if any of the array values matches.

To simplify handling of multilevel and binary switches, metrics:level can
both be queried with the values 'on','off', true and false. (ie.
['metrics:level','=','on'] would also find multilevel switches with level >= 1

## processDeviceList

Processes a callback for each device in the list of device

```javascript
self.processDeviceList(self.config.devices,function(vDev) {
    // Do something
});
```

## parseTime

Parses a string in the format HH:MM and returns a Date object fot the current day

## checkPeriod

Expects two time strings (HH:MM) marking a period, and calculates if the
period is currently matching.

# Configuration

No configuration required

# Virtual Devices

No virtual device is created

# Events

All virtual devices will emit a modify:metrics:level event if the
metrics:level value was modified. In contrast to change this event will only
be triggered if metrics:level was set with a different value.

# Installation

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is
required. If you want to test the latest pre-releases use 'k1_beta' as
app store access token.

For developers and users of older Zway versions installation via git is
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-BaseModule.git BaseModule --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/BaseModule
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

Even though this module has no user-facing functionality on its own it has
to be instantiated via Apps > Local Apps before it can be used by other modules.

# License

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.11 2019 Apr 26, Fri
    * Also handle null values in comparison
    * Improve handling of metrics:target
    * Improve compare value fallbacks
    * Log time parse errors

1.10 2017 Jan 22, Sun
    * Improved period search: Close gaps
    * Update documentation
    * Improved device search to allow multiple search values
    * Internal changes

1.09 2016 Jul 03, Sun
    * Cleanup and add fallbacks
    * Check for undefinedness
    * Use correct key to access last update time

1.08 2016 Apr 20, Wed
    * Add processDevices helper
    * Improve sanity checks
    * Alert on unlikely level changes
    * Fix regexp comparison
    * Fix device search
    * Update time parser to handle HH:MM:SS am/pm
    * Refactor comparison
    * Fix getDevice by ID
    * Update presence probeType

1.07 2016 Mar 18, Fri
    * Add base image path

1.06 2016 Mar 08, Tue
    * Fix superfluous argument to vDev.set
    * Also accept IDs for getDevice and getDevice value
    * processDeviceList now also accepts list of object or single object/id
    * Update installation howto

1.05 2016 Feb 20, Sat
    * Fix module.json

1.04 2016 Feb 15, Mon
    * Add compare helper
    * Add doc link

1.03 2016 Jan 08, Fri
    * Rename title/description key to make module work in appstore
    * Clarify documentation
    * modify:metrics:level events
    * Update documentation
    * Emit event on real change
    * Fix period parser

1.02 2016 Jan 02, Sat
    * Clarify log messages
    * Coerce for comparison
    * Update homepage
    * Use constructor.name instead of getName to fetch class name
    * Update readme

1.01 2015 Dec 29, Tue
    * Fix time period calculation
    * German translation
    * Delayed lastLevel set
    * Fix duplicate error message

1.00 2015 Dec 26, Sat
    * Initial release


index.js
/*** BaseModule Z-Way HA module *******************************************

Version: 1.11
(c) Maroš Kollár, 2015-2017
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    This module provides many helpful functions for zway automation module
    developers.

******************************************************************************/

/*jshint -W058 */

function BaseModule (id, controller) {
    // Call superconstructor first (AutomationModule)
    BaseModule.super_.call(this, id, controller);

    this.langFile       = undefined;
}

inherits(BaseModule, AutomationModule);

_module = BaseModule;

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

BaseModule.prototype.init = function (config) {
    BaseModule.super_.prototype.init.call(this, config);
    var self = this;

    self.langFile   = self.controller.loadModuleLang(self.constructor.name);
    self.imagePath  = '/ZAutomation/api/v1/load/modulemedia/'+self.constructor.name;

    // Only for BaseModule instance
    if (self.constructor.name === 'BaseModule') {
        self.log('Init callbacks');
        self.callbackBase = _.bind(self.handleLevelModification,self);
        self.controller.devices.on('change:metrics:level',self.callbackBase);
    }
};

BaseModule.prototype.stop = function () {
    var self = this;

    // Only for BaseModule instance
    if (self.constructor.name === 'BaseModule') {
        self.controller.devices.off('change:metrics:level',self.callbackBase);
        self.callbackBase = undefined;
    }

    BaseModule.super_.prototype.stop.call(this);
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

BaseModule.prototype.presenceModes = ['home','night','away','vacation'];

BaseModule.prototype.handleLevelModification = function(vDev) {
    var self = this;

    var lastLevel           = vDev.get('metrics:lastLevel');
    var newLevel            = vDev.get('metrics:level');
    var deviceType          = vDev.get('deviceType');
    var lastUpdate          = vDev.get('updateTime');
    var modificationTime    = Math.floor(new Date().getTime() / 1000);

    // No lastlevel - set it for the first time
    if (typeof(lastLevel) === 'undefined') {
        vDev.set('metrics:modificationTime',modificationTime,{ silent: true });
        vDev.set('metrics:lastLevel',newLevel,{ silent: true });
        return;
    }

    // Not changed
    if (lastLevel == newLevel) return;

    // Warn on big level changes - TODO maybe deny too big changes
    if (deviceType === 'sensorMultilevel') {
        var diff = Math.abs(lastLevel-newLevel);
        var probeType = vDev.get('metrics:probeType');
        // Ignore diff for changes of more than 4 hours
        if (modificationTime-lastUpdate > (4*60*60)) {
            diff = 0;
        }
        if (
                (probeType == 'luminosity' && (diff > 250 || newLevel > 1000 || newLevel < 0)) ||
                (probeType == 'temperature' && (diff > 10 || newLevel > 50 || newLevel < -30)) ||
                (probeType == 'humidity' && (diff > 20 || newLevel > 100 || newLevel < 5)) ||
                (probeType == 'ultraviolet' && (diff > 4 || newLevel > 15 || newLevel < -1))
            ) {
            self.error('Unlikely '+probeType+' level change from '+lastLevel+' to '+newLevel+' for '+vDev.id);
        }
    }

    // Run delayed, in order not to delay current processing
    setTimeout(function() {
        // Set modificationTime
        self.log('Set lastLevel to '+newLevel+' for '+vDev.id+' (was '+lastLevel+')');
        vDev.set('metrics:modificationTime',modificationTime,{ silent: true });
        vDev.set('metrics:lastLevel',newLevel,{ silent: true });
    },1);

    // Bind to modify:metrics:level to get real changes
    self.controller.devices.emit('modify:metrics:level',vDev,'metrics:level');
    vDev.emit('modify:metrics:level',vDev,'metrics:level');
};

/* Log helper functions */

BaseModule.prototype.log = function(message) {
    if (undefined === message) return;
    console.log('['+this.constructor.name+'-'+this.id+'] '+message);
};

BaseModule.prototype.error = function(message) {
    if (undefined === message) message = 'An unknown error occured';
    var error = new Error(message);
    console.error('['+this.constructor.name+'_'+this.id+'] '+error.stack);
};

/* Presence helper functions */

BaseModule.prototype.getPresenceBoolean = function() {
    var self = this;

    var value = self.getDeviceValue([
        ['probeType','=','presence']
    ]);

    if (typeof(value) === 'string' && value === 'on') {
        return true;
    } else if (typeof(value) === 'undefined') {
        self.error('Could not find presence device');
        return true; // Fallback
    }
    return false;
};

BaseModule.prototype.getPresenceMode = function() {
    var self = this;

    var value = self.getDeviceValue([
        ['probeType','=','presence']
    ],'metrics:mode');

    if (typeof(value) === 'undefined') {
        self.error('Could not find presence device');
        return 'home'; // Fallback
    }

    return value;
};

/* Device helper functions */

BaseModule.prototype.processDeviceList = function(devices,callback) {
    var self = this;
    if (! _.isFunction(callback)) {
        self.error('Invalid callback for processDeviceList');
        return;
    }

    if (_.isUndefined(devices) === 'undefined') {
        return;
    } else if (! _.isArray(devices)) {
        devices = [ devices ];
    }

    _.each(devices,function(device) {
        var vDev;
        if (_.isString(device)) {
            vDev = self.controller.devices.get(device);

        } else if (_.isObject(device)) {
            vDev = device;
        }
        if (_.isNull(vDev) || _.isUndefined(vDev)) {
            self.error('Device not found '+device);
            return;
        }
        callback(vDev);
    });
};

BaseModule.prototype.compareDevice = function(vDev,criterias) {
    var self = this;

    var match;
    _.each(criterias,function(criteria) {
        // Matched false - skip the rest
        if (match === false) {
            return;
        // Device ID
        } else if (_.isString(criteria)) {
            if (criteria !== vDev.id) {
                match = false;
            }
        // Comparison array
        } else  if (_.isArray(criteria)) {
            // eg. ['metrics:level','=','on']
            var matchKey        = criteria[0];
            var matchComparison = criteria[1];
            var matchList       = criteria[2];
            var compareValue;
            // Get value
            if (matchKey === 'zwaveId') {
                var result = vDev.id.match(/^ZWayVDev_zway_(\d+)-\d+-/m);
                if (result === null) {
                    match = false;
                    return;
                }
                compareValue = parseInt(result[1],10);
            } else {
                compareValue = vDev.get(matchKey);
                // Fall back for metrics:level if no target is set
                if (matchKey === 'metrics:target'
                    && (typeof(compareValue) === 'undefined' || compareValue === null)) {
                    compareValue = vDev.get('metrics:level');
                }
            }

            if (typeof(compareValue) === 'undefined' || compareValue === null) {
                return false;
            }

            if (!_.isArray(matchList)) {
                matchList = [ matchList ];
            }

            var matchFirst = _.find(matchList,function(matchValue) {
                // Comparison is array
                if (_.isArray(compareValue)) {
                    var hasMatch = (compareValue.indexOf(matchValue) === -1) ? false:true;
                    return ((matchComparison === '!=' && !hasMatch) || (matchComparison === '=' && hasMatch)) ? true:false;
                // Comparison is scalar
                } else {
                    if (matchKey === 'metrics:level' || matchKey === 'metrics:target') {
                        if (vDev.get('deviceType') === 'switchBinary') {
                            if (typeof(matchValue) === 'number') {
                                compareValue = (compareValue === 'on' ? 255:0);
                            } else if (typeof(matchValue) === 'boolean') {
                                compareValue = (compareValue === 'on' ? true:false);
                            }
                        } else if (vDev.get('deviceType') === 'switchMultilevel') {
                            if (typeof(matchValue) === 'string') {
                                compareValue = (compareValue > 0 ? 'on':'off');
                            } else if (typeof(matchValue) === 'boolean') {
                                compareValue = (compareValue > 0 ? true:false);
                            }
                        }
                    }
                    return self.compare(compareValue,matchComparison,matchValue);
                }
            });

            match = ! _.isUndefined(matchFirst);
        } else {
            self.error('Invalid device comparison');
        }
    });
    if (typeof(match) === 'boolean') {
        return match;
    }
    return false;
};

BaseModule.prototype.processDevices = function(criterias,callback) {
    var self = this;

    if (! _.isFunction(callback)) {
        self.error('Invalid callback for processDevices');
        return;
    }

    self.controller.devices.each(function(vDev) {
        var match = self.compareDevice(vDev,criterias);
        if(match === true) {
            callback(vDev);
        }
    });
};

BaseModule.prototype.getDevices = function(criterias) {
    var self = this;

    var devices = [];
    self.controller.devices.each(function(vDev) {
        var match = self.compareDevice(vDev,criterias);
        if(match === true) {
            devices.push(vDev);
        }
    });
    return devices;
};

BaseModule.prototype.getDevice = function(criterias) {
    var self = this;

    var device;
    if (typeof(criterias) === 'string') {
        device = self.controller.devices.get(criterias);
    } else if (_.isArray(criterias)) {
        self.controller.devices.each(function(vDev) {
            if (typeof(device) === 'undefined') {
                var match = self.compareDevice(vDev,criterias);
                if(match === true) {
                    device = vDev;
                }
            }
        });
    }
    if (device === null) {
        return undefined;
    } else {
        return device;
    }
};

BaseModule.prototype.getDeviceValue = function(criterias,key) {
    var self = this;
    key = key || 'metrics:level';
    var device = self.getDevice(criterias);
    if (typeof(device) === 'undefined') {
        self.error('Could not find device');
    } else {
        return device.get(key);
    }
};

BaseModule.prototype.performCommandDevices = function(criterias,command,args,auto) {
    var self = this;
    args = args || {};

    var devices = [];
    self.controller.devices.each(function(vDev) {
        var match = self.compareDevice(vDev,criterias);

        if(match === true) {
            devices.push(vDev);
            if (typeof(command) !== 'undefined') {
                setTimeout(function() {
                    vDev.performCommand(command,args);
                },1);
            }
            if (typeof(auto) === 'boolean') {
                vDev.set('metrics:auto',auto);
            }
        }
    });
    return devices;
};

/* Time helper functions */

BaseModule.prototype.parseTime = function(timeString) {
    if (typeof(timeString) === 'undefined') {
        return;
    }

    var match = timeString.match(/^(\d{1,2}):(\d{1,2})(?::(\d{1,2}))?\s*(am|pm)?$/);
    if (!match) {
        this.log('Could not parse time: '+timeString);
        return;
    }
    var hour        = parseInt(match[1],10);
    var minute      = parseInt(match[2],10);
    var second      = parseInt(match[3] || 0,10);
    var ampm        = match[4];
    if (ampm === 'pm' && hour < 12) {
        hour = hour + 12;
    }
    var dateCalc    = new Date();
    dateCalc.setHours(hour, minute,second,0);

    return dateCalc;
};

BaseModule.prototype.checkPeriod = function(timeFrom,timeTo) {
    var self = this;

    var dateNow = new Date();

    // Check from/to time
    if (typeof(timeFrom) === 'string') {
        timeFrom = self.parseTime(timeFrom);
    }
    if (typeof(timeTo) === 'string') {
        timeTo = self.parseTime(timeTo);
    }

    if (typeof(timeFrom) === 'undefined'
        || typeof(timeTo) === 'undefined') {
        return true;
    }

    // Period over midnight
    if (timeFrom >= timeTo) {
        if (timeTo <= dateNow) {
            var toHour   = timeTo.getHours();
            var toMinute = timeTo.getMinutes();
            timeTo.setHours(toHour + 24);
            // Now fix time jump on DST
            timeTo.setHours(toHour,toMinute);
        } else {
            var fromHour     = timeFrom.getHours();
            var fromMinute   = timeFrom.getMinutes();
            timeFrom.setHours(fromHour - 24);
            // Now fix time jump on DST
            timeFrom.setHours(fromHour,fromMinute);
        }
    }

    if (timeFrom > dateNow || dateNow > timeTo) {
        return false;
    }

    self.log('Match '+timeFrom+'-'+timeTo);
    return true;
};

BaseModule.prototype.compare = function (val1, op, val2) {
    // match = self.compare(compareValue,matchComparison,matchValue);
    if (op === "=") {
        if (_.isRegExp(val2) && _.isString(val1)) {
            return val1.match(val2) !== null;
        } else {
            return val1 === val2;
        }
    } else if (op === "!=") {
        if (_.isRegExp(val2) && _.isString(val1)) {
            return val1.match(val2) === null;
        } else {
            return val2 !== val1;
        }
    } else if (op === ">") {
        return val1 > val2;
    } else if (op === "<") {
        return val1 < val2;
    } else if (op === ">=") {
        return val1 >= val2;
    } else if (op === "<=") {
        return val1 <= val2;
    } else {
        console.error('Invalid comparison operator '+op);
    }

    return null; // error!!
};

//----------------------------------------------------------------------------
//--- Timeout manager
//----------------------------------------------------------------------------

function TimeoutManager(scope) {
    var self        = this;
    self.scope      = scope;
    self.timeouts   = {};
}
TimeoutManager.prototype.timeouts = {};
TimeoutManager.prototype.running  = function(id) {
    var self   = this;
    if (typeof(self.timeouts[id]) === "undefined") {
        return false;
    }
    return self.timeouts[id].cleared ? false:true;
};
TimeoutManager.prototype.add      = function(id,fn,interval) {
    var self   = this;

    // Do not overwrite existing timer
    if (typeof(self.timeouts[id]) !== "undefined"
        && self.timeouts[id].cleared === false) {
        throw('Timeout "'+id+'" is already active');
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope,self.scope);
    self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
    return self.timeouts[id];
};
TimeoutManager.prototype.replace  = function(id,fn,interval) {
    var self   = this;

    // Clear existing timer
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope,self.scope);
    self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
    return self.timeouts[id];
};
TimeoutManager.prototype.get    = function(id) {
    return this.timeouts[id];
};
TimeoutManager.prototype.clear    = function(id) {
    var self   = this;
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
        delete self.timeouts[id];
        return true;
    }
    return false;
};
TimeoutManager.prototype.clearAll = function() {
    var self   = this;
    for (var id in self.timeouts) {
        if (self.timeouts.hasOwnProperty(id)
            && typeof(self.timeouts[id]) !== "undefined") {
            self.timeouts[id].clear();
        }
    }
    self.timeouts = {};
};

//----------------------------------------------------------------------------
//--- Timeout class
//----------------------------------------------------------------------------

function Timeout(scope,fn,interval) {
    var self   = this;
    var args   = new Array(arguments.length-3);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+3];
    }
    args.unshift(scope);
    fn          = fn || function() {};

    self.fn     = Function.prototype.bind.apply(fn,args);
    self.id     = setTimeout(self.run.bind(self),interval);
    self.cleared= false;
    //Register timeout by name?
}

Timeout.prototype.id       = undefined;
Timeout.prototype.cleared  = false;
Timeout.prototype.fn       = undefined;
Timeout.prototype.run      = function() {
    this.clear();
    this.fn();
};
Timeout.prototype.clear    = function() {
    var self   = this;
    if (typeof(self.id) !== 'undefined' && self.cleared === false) {
        clearTimeout(self.id);
    }
    self.cleared = true;
    self.id = undefined;
};
